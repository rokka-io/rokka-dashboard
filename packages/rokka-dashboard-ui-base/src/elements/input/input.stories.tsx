import { action } from '@storybook/addon-actions';
import { boolean, number, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { BooleanInput, NumberInput, TextInput } from './input';

storiesOf('Elements / Forms', module)
  .add('TextInput', () => (
    <TextInput
      name="textInputStory"
      value={text('textInput', 'test')}
      onChange={action('onChange')}
      placeholder={text('text', 'A placeholder')}
      disabled={boolean('disabled', false)}
    />
  ))
  .add('BooleanInput', () => (
    <BooleanInput
      name="booleanInputStory"
      value={boolean('booleanInput', false)}
      onChange={action('onChange')}
      disabled={boolean('disabled', false)}
    />
  ))
  .add('NumberInput', () => (
    <NumberInput
      name="numberInputStory"
      value={number('numberInput', 10)}
      onChange={action('onChange')}
      disabled={boolean('disabled', false)}
    />
  ));
