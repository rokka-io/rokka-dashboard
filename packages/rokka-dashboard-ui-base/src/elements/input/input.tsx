import React, { ReactElement, SyntheticEvent } from 'react';
import styled from 'styled-components';
import { colors } from '../../identity/colors/colors';
import { spaces } from '../../identity/spaces/spaces';
import { fonts } from '../../identity/typography';

interface InputProps<T> {
  /** input name */
  name: string;
  /** input value */
  value: T;
  /** onChange is fired when the input changes */
  onChange: (name: string, value: T) => void;
  /** deserialize from T to a string or null. Used for rawValue. */
  deserialize: (value: T) => string | null;
  /** serialize to T from target */
  serialize: (target: HTMLInputElement) => T;
  /** render prop */
  children: (
    props: {
      value: T;
      rawValue: string | null;
      onChange: (evt: SyntheticEvent) => void;
    }
  ) => ReactElement<{}>;
}

function Input<T>({ name, value, onChange, deserialize, serialize, children }: InputProps<T>) {
  return children({
    value,
    rawValue: deserialize(value),
    onChange: (evt: SyntheticEvent) => {
      evt.preventDefault();
      onChange(name, serialize(evt.target as HTMLInputElement));
    }
  });
}

interface TextInputProps {
  /** input name */
  name: string;
  /** input value */
  value: string;
  /** onChange fired when input value changes */
  onChange: (name: string, value: string) => void;
  /** placeholder for input */
  placeholder?: string;
  /** input type, default text */
  type?: string;
  /** is input disabled */
  disabled?: boolean;
  /** className mainly used for styled components */
  className?: string;
}

export function TextInput({ name, value, onChange, type = 'text', ...props }: TextInputProps) {
  return (
    <Input name={name} value={value} onChange={onChange} serialize={r => r.value} deserialize={v => v}>
      {p => <StyledInput name={name} value={p.value} onChange={p.onChange} type={type} {...props} />}
    </Input>
  );
}

interface BooleanInputProps {
  /** input name */
  name: string;
  /** input value */
  value: boolean;
  /** onChange fired when input value changes */
  onChange: (name: string, value: boolean) => void;
  /** is input disabled */
  disabled?: boolean;
  /** className mainly used for styled components */
  className?: string;
}

export function BooleanInput({ name, value, onChange, ...props }: BooleanInputProps) {
  return (
    <Input
      name={name}
      value={value}
      onChange={onChange}
      serialize={r => r.checked}
      deserialize={v => (v ? 'true' : 'false')}
    >
      {p => <StyledInput checked={p.value} onChange={p.onChange} type="checkbox" {...props} />}
    </Input>
  );
}

interface NumberInputProps {
  /** input name */
  name: string;
  /** input value */
  value: number;
  onChange: (name: string, value: number) => void;
  /** is input disabled */
  disabled?: boolean;
  /** className mainly used for styled components */
  className?: string;
}

export function NumberInput({ name, value, onChange, ...props }: NumberInputProps) {
  return (
    <Input
      name={name}
      value={value}
      onChange={onChange}
      serialize={r => Number.parseInt(r.value, 10)}
      deserialize={v => `${v}`}
    >
      {p => <StyledInput value={p.value} onChange={p.onChange} type="number" {...props} />}
    </Input>
  );
}

const StyledInput = styled.input`
  font-family: ${fonts.Families.brand};
  font-size: ${fonts.Sizes.small};
  width: 100%;
  height: 38px;
  line-height: 38px;
  border: 1px solid ${colors.gray.light};
  padding: 0 ${spaces.small};
  outline: none;

  :focus {
    transition: border 0.4s ease;
    border-color: ${colors.gray.darker};
  }

  :disabled {
    color: ${colors.gray.darkest};
    background: ${colors.gray.lighter};
  }
`;
