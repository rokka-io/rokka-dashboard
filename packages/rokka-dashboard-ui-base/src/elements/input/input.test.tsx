import { mount } from 'enzyme';
import React from 'react';

import { BooleanInput, NumberInput, TextInput } from './input';

describe('Elements', () => {
  describe('Input', () => {
    it('renders a TextInput correctly.', () => {
      const onChange = jest.fn();
      const component = mount(<TextInput name="testname" value="Test" onChange={onChange} />);

      expect(component).toMatchSnapshot();

      component.find('input').simulate('change', {
        target: {
          name: 'testname',
          value: 'another test'
        }
      });

      expect(onChange).toHaveBeenCalledWith('testname', 'another test');
    });
    it('renders a BooleanInput correctly.', () => {
      const onChange = jest.fn();
      const component = mount(<BooleanInput name="testname" value={true} onChange={onChange} />);

      expect(component).toMatchSnapshot();

      component.find('input').simulate('change', {
        target: {
          name: 'testname',
          checked: false
        }
      });

      expect(onChange).toHaveBeenCalledWith('testname', false);
    });

    it('renders a NumberInput correctly.', () => {
      const onChange = jest.fn();
      const component = mount(<NumberInput name="testname" value={10} onChange={onChange} />);

      expect(component).toMatchSnapshot();

      component.find('input').simulate('change', {
        target: {
          name: 'testname',
          value: '20'
        }
      });

      expect(onChange).toHaveBeenCalledWith('testname', 20);
    });
  });
});
