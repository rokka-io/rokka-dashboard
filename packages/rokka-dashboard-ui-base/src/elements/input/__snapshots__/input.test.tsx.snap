// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements Input renders a BooleanInput correctly. 1`] = `
<BooleanInput
  name="testname"
  onChange={[MockFunction]}
  value={true}
>
  <Input
    deserialize={[Function]}
    name="testname"
    onChange={[MockFunction]}
    serialize={[Function]}
    value={true}
  >
    <styled.input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    >
      <StyledComponent
        checked={true}
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": true,
              "lastClassName": "hPvubn",
              "rules": Array [
                "
  font-family: ",
                "\\"Liip Etica Book\\", sans-serif",
                ";
  font-size: ",
                "0.875rem",
                ";
  width: 100%;
  height: 38px;
  line-height: 38px;
  border: 1px solid ",
                "#DDDDD8",
                ";
  padding: 0 ",
                "0.5rem",
                ";
  outline: none;

  :focus {
    transition: border 0.4s ease;
    border-color: ",
                "#A5A59F",
                ";
  }

  :disabled {
    color: ",
                "#898A81",
                ";
    background: ",
                "#EBEBE6",
                ";
  }
",
              ],
            },
            "displayName": "styled.input",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "input",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        onChange={[Function]}
        type="checkbox"
      >
        <input
          checked={true}
          className="sc-bdVaJa hPvubn"
          onChange={[Function]}
          type="checkbox"
        />
      </StyledComponent>
    </styled.input>
  </Input>
</BooleanInput>
`;

exports[`Elements Input renders a NumberInput correctly. 1`] = `
<NumberInput
  name="testname"
  onChange={[MockFunction]}
  value={10}
>
  <Input
    deserialize={[Function]}
    name="testname"
    onChange={[MockFunction]}
    serialize={[Function]}
    value={10}
  >
    <styled.input
      onChange={[Function]}
      type="number"
      value={10}
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": true,
              "lastClassName": "hPvubn",
              "rules": Array [
                "
  font-family: ",
                "\\"Liip Etica Book\\", sans-serif",
                ";
  font-size: ",
                "0.875rem",
                ";
  width: 100%;
  height: 38px;
  line-height: 38px;
  border: 1px solid ",
                "#DDDDD8",
                ";
  padding: 0 ",
                "0.5rem",
                ";
  outline: none;

  :focus {
    transition: border 0.4s ease;
    border-color: ",
                "#A5A59F",
                ";
  }

  :disabled {
    color: ",
                "#898A81",
                ";
    background: ",
                "#EBEBE6",
                ";
  }
",
              ],
            },
            "displayName": "styled.input",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "input",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        onChange={[Function]}
        type="number"
        value={10}
      >
        <input
          className="sc-bdVaJa hPvubn"
          onChange={[Function]}
          type="number"
          value={10}
        />
      </StyledComponent>
    </styled.input>
  </Input>
</NumberInput>
`;

exports[`Elements Input renders a TextInput correctly. 1`] = `
<TextInput
  name="testname"
  onChange={[MockFunction]}
  value="Test"
>
  <Input
    deserialize={[Function]}
    name="testname"
    onChange={[MockFunction]}
    serialize={[Function]}
    value="Test"
  >
    <styled.input
      name="testname"
      onChange={[Function]}
      type="text"
      value="Test"
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": true,
              "lastClassName": "hPvubn",
              "rules": Array [
                "
  font-family: ",
                "\\"Liip Etica Book\\", sans-serif",
                ";
  font-size: ",
                "0.875rem",
                ";
  width: 100%;
  height: 38px;
  line-height: 38px;
  border: 1px solid ",
                "#DDDDD8",
                ";
  padding: 0 ",
                "0.5rem",
                ";
  outline: none;

  :focus {
    transition: border 0.4s ease;
    border-color: ",
                "#A5A59F",
                ";
  }

  :disabled {
    color: ",
                "#898A81",
                ";
    background: ",
                "#EBEBE6",
                ";
  }
",
              ],
            },
            "displayName": "styled.input",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "input",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        name="testname"
        onChange={[Function]}
        type="text"
        value="Test"
      >
        <input
          className="sc-bdVaJa hPvubn"
          name="testname"
          onChange={[Function]}
          type="text"
          value="Test"
        />
      </StyledComponent>
    </styled.input>
  </Input>
</TextInput>
`;
